libs/isograph-react/src/core/cache.ts

```diff
@@ -412,11 +438,26 @@ function normalizeDataIntoRecord(
     }
   }
   if (recordHasBeenUpdated) {
-    mutableEncounteredIds.add(targetParentRecordLink.__link);
+    let encounteredRecordsIds = insertIfNotExists(
+      mutableEncounteredIds,
+      targetParentRecordLink.__typename,
+      new Set(),
+    );
+
+    encounteredRecordsIds.add(targetParentRecordLink.__link);
   }
   return recordHasBeenUpdated;
 }
 
+export function insertIfNotExists<K, V>(map: Map<K, V>, key: K, value: V) {
```

Comment:
nit: can we call this `getOrCreate` and rename `value` to `defaultValue`? That might be more intuitive.

Also, it would be more performant if we didn't create a new `Set` every time this was called. Can we make map `Map<K, Set<V>>`, not have a `value` param, and say `result = new Set()` below?